flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java

```diff
@@ -298,19 +306,24 @@ void serializeKeyedStateHandle(KeyedStateHandle stateHandle, DataOutputStream do
             if (stateHandle instanceof KeyGroupsSavepointStateHandle) {
                 dos.writeByte(SAVEPOINT_KEY_GROUPS_HANDLE);
             } else {
-                dos.writeByte(KEY_GROUPS_HANDLE);
+                dos.writeByte(KEY_GROUPS_HANDLE_V2);
             }
             dos.writeInt(keyGroupsStateHandle.getKeyGroupRange().getStartKeyGroup());
             dos.writeInt(keyGroupsStateHandle.getKeyGroupRange().getNumberOfKeyGroups());
             for (int keyGroup : keyGroupsStateHandle.getKeyGroupRange()) {
                 dos.writeLong(keyGroupsStateHandle.getOffsetForKeyGroup(keyGroup));
             }
             serializeStreamStateHandle(keyGroupsStateHandle.getDelegateStateHandle(), dos);
+
+            // savepoint state handle would not need to persist state handle id out.
+            if (!(stateHandle instanceof KeyGroupsSavepointStateHandle)) {
+                dos.writeUTF(stateHandle.getStateHandleId().toString());
+            }
```

Comment:
I see, another reason that I don't want to write the state handle id is because that I have to introduce another `SAVEPOINT_KEY_GROUPS_HANDLE_V2`. I just don't want introduce to much V2 version flag to disable restoring savepoint of Flink-1.15 to Flink-1.14 (forward compatibilit), although this is not offcially supported.

Reply:
Indeed, that's an additional complexity.
Thanks for clarifying.